### PR TITLE
Rl fix

### PR DESCRIPTION
**Pull request by liruipeng**
@rfalgout @ulrikeyang 
Several bug fixes on GPU build, PILUT compile errors with gcc, and interfacing with PETSc

1. It seems that GCC compilers (new versions, see #126) get confused by something in PILUT like 
```
typedef struct {
HYPRE_Int mype;
...
};
#define mype (global->mype)
```
where `mype` appears in both `#define` and the struct. Seems to have been fixed by changing the names in the struct.

2. `atomicAdd`, used in sparse matrix-matrix, was supported for `double` only after `CUDA_ARCH` 60. Add a workaround for older CUDA's (see https://github.com/hypre-space/hypre/issues/125)

3. In hypre 2.19, `hypre_AuxParCSRMatrixInitialize` was renamed to its `_v2` interface in with an additional `memory_location` argument. The original function is still needed by PETSc. Had to add it back with default `memory_location`. 

4. There was an issue [reported by Holger Jones, issue1772] from PETSc where `hypre_IJMatrixSetDiagOffdSizesParCSR` https://github.com/hypre-space/hypre/blob/f875c2a38b52d10264ecfa5a0a4114eb659510b0/src/IJ_mv/IJMatrix_parcsr.c#L197 was called, which set internal flag `hypre_AuxParCSRMatrixNeedAux(aux_matrix)` https://github.com/hypre-space/hypre/blob/f875c2a38b52d10264ecfa5a0a4114eb659510b0/src/IJ_mv/IJMatrix_parcsr.c#L237 to `0`, but this flag was manipulated in PETSc after (set back to `1`) for some reason. It was also related to our code refactor in 2.19 that broke PETSc.

### CODE REVIEW

File: src/distributed_ls/pilut/DistributedMatrixPilutSolver.h [hunk 1]
```diff
@@ -33,23 +33,23 @@
 typedef struct
 {
 MPI_Comm hypre_MPI_communicator;
-HYPRE_Int mype, npes;
+HYPRE_Int _mype, _npes;
```
Comment (rfalgout) [inline src/distributed_ls/pilut/DistributedMatrixPilutSolver.h:36]: @liruipeng What's the reason for the underscores?

File: src/utilities/_hypre_utilities.h [hunk 2]
```diff
@@ -1455,6 +1456,26 @@ hypre_int hypre_cuda_get_grid_warp_id()
           hypre_cuda_get_warp_id<bdim>();
 }
 
+#if defined(__CUDA_ARCH__) && __CUDA_ARCH__ < 600
```
Comment (rfalgout) [inline src/utilities/_hypre_utilities.h:1459]: @liruipeng What's going on here?  Where is atomicAdd even called?

**Review by rfalgout — COMMENTED** [review]
Hi @liruipeng ,

Could you provide a bit more description about what's going on with this pull request.  I'm having difficulty seeing what it is fixing, especially with respect to PETSc.  Thanks!

-Rob

File: src/distributed_ls/pilut/DistributedMatrixPilutSolver.h [hunk 3]
```diff
@@ -78,31 +78,31 @@ HYPRE_Int Ul_timer;
 
 /* DEFINES for global variables */
 #define pilut_comm (globals->hypre_MPI_communicator)
-#define mype (globals->mype)
```
Comment (liruipeng) [inline src/distributed_ls/pilut/DistributedMatrixPilutSolver.h:81]: @rfalgout My guess is GCC compilers (new versions, see https://github.com/hypre-space/hypre/issues/126#issue-625738732) got confused from these lines (starting from 81). Must have something to do with name conflicts in, e.g., `#define mype (global->mype)` and `mype` in the struct. Seems working with the new changes.

**Review by rfalgout — APPROVED** [review]
Looks good, @liruipeng .  Thanks!

**Review by ulrikeyang — APPROVED** [review]
Looks good

**Pull request merged**